packages/network-controller/src/NetworkController.test.ts

```diff
@@ -3722,6 +3724,108 @@ describe('NetworkController', () => {
       });
     });
   });
+
+  describe('Network Configurations', () => {
+    it('should add a network configuration', async () => {
+      const messenger = buildMessenger();
+      await withController({ messenger }, async ({ controller }) => {
+        const rpcUrlNetwork = {
+          chainId: undefined,
+          nickname: 'RPC',
+          rpcPrefs: undefined,
+          rpcUrl: 'rpc_url',
+          ticker: 'RPC',
+        };
+        const localhostNetwork = {
+          chainId: undefined,
+          nickname: undefined,
+          rpcPrefs: undefined,
+          rpcUrl: 'http://localhost:8545',
+          ticker: 'LOCAL',
+        };
+        controller.upsertNetworkConfiguration(rpcUrlNetwork);
+        controller.upsertNetworkConfiguration(localhostNetwork);
+
+        expect(
+          Object.values(controller.state.networkConfigurations),
+        ).toStrictEqual(
+          expect.arrayContaining([rpcUrlNetwork, localhostNetwork]),
+        );
+        controller.upsertNetworkConfiguration({ rpcUrl: 'rpc_url' });
+
+        expect(
+          Object.values(controller.state.networkConfigurations),
+        ).toStrictEqual(
+          expect.arrayContaining([
+            localhostNetwork,
+            { ...rpcUrlNetwork, nickname: undefined, ticker: undefined },
+          ]),
+        );
+      });
+    });
+
+    it('should update a network configuration when the configuration being added has an rpcURL that matches an existing configuration', async () => {
+      const messenger = buildMessenger();
+      await withController({ messenger }, async ({ controller }) => {
+        controller.upsertNetworkConfiguration({
+          rpcUrl: 'test_rpc_url',
+          ticker: 'old_rpc_ticker',
+          nickname: 'old_rpc_nickname',
+        });
+
+        expect(
+          Object.values(controller.state.networkConfigurations),
+        ).toStrictEqual([
+          expect.objectContaining({
+            rpcUrl: 'test_rpc_url',
+            nickname: 'old_rpc_nickname',
+            ticker: 'old_rpc_ticker',
+            rpcPrefs: undefined,
+            chainId: undefined,
+          }),
+        ]);
+
+        controller.upsertNetworkConfiguration({
+          rpcUrl: 'test_rpc_url',
+          ticker: 'new_rpc_ticker',
+          nickname: 'new_rpc_nickname',
+        });
+
+        expect(
+          Object.values(controller.state.networkConfigurations),
+        ).toStrictEqual([
+          expect.objectContaining({
+            rpcUrl: 'test_rpc_url',
+            nickname: 'new_rpc_nickname',
+            ticker: 'new_rpc_ticker',
+            rpcPrefs: undefined,
+            chainId: undefined,
+          }),
+        ]);
+      });
```

Comment:
Similar thing — I wonder if we need to call `upsertNetworkConfiguration` twice here. If the goal is to test that if we already have a network configuration in state and we call it again with the same `rpcURL` the existing one gets updated, we could add that existing one via state directly without going through `upsertNetworkConfiguration`. We also don't really need to test that the state is updated when we call `upsertNetworkConfiguration`, as that's handled in another test. What do you think about:

``` suggestion
      await withController(
        {
          state: {
            networkConfigurations: [
              {
                rpcUrl: 'test_rpc_url',
                ticker: 'old_rpc_ticker',
                nickname: 'old_rpc_nickname',
                rpcPrefs: undefined,
                chainId: undefined,
              }
            ]
          }
        },
        async ({ controller }) => {
          controller.upsertNetworkConfiguration({
            rpcUrl: 'test_rpc_url',
            ticker: 'new_rpc_ticker',
            nickname: 'new_rpc_nickname',
          });

          expect(
            Object.values(controller.state.networkConfigurations),
          ).toStrictEqual([
            expect.objectContaining({
              rpcUrl: 'test_rpc_url',
              nickname: 'new_rpc_nickname',
              ticker: 'new_rpc_ticker',
              rpcPrefs: undefined,
              chainId: undefined,
            }),
          ]);
        }
      );
```